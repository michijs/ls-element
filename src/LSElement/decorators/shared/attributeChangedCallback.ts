import { LSCustomElement } from '../../types';
import { getAttributeValue } from '../../utils/getAttribute';

export function attributeChangedCallback(self: LSCustomElement, name: string, oldValue, newValue) {
  if (newValue != oldValue) {
    self.componentWillReceiveAttribute?.(name, newValue, oldValue);
    self[name] = getAttributeValue(newValue);
  }
}