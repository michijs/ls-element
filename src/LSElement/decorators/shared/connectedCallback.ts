import type { LSCustomElement } from '../../types';
import { addReflectedAttributes } from '../../properties/addReflectedAttributes';
import { executeFirstRender } from '../../DOM/executeFirstRender';

export function connectedCallback(self: LSCustomElement) {//Children and attributes should be managed after constructor
  addReflectedAttributes(self);
  if (!self.ls.alreadyRendered) {
    self.componentWillMount?.();
    executeFirstRender(self);
    self.ls.alreadyRendered = true;
    self.componentDidMount?.();
  }
}