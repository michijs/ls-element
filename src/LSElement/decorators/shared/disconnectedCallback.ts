import { removeStoredAttributes } from '../../properties/removeStoredAttributes';
import type { LSCustomElement } from '../../types';

export function disconnectedCallback(self: LSCustomElement) {
  if (self.parentNode === null) {//TODO: search a better way to validate if element does not exists anymore
    removeStoredAttributes(self);
    self.componentDidUnmount?.();
  }
}