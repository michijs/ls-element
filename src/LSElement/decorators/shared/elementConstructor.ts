import type { LSCustomElement, ShadowOption, OtherShadowOptions } from '../../types';
import { addStores } from '../../properties/addStores';
import { initStore } from '../../properties/initStore';
import { addAttributes } from '../../properties/addAttributes';
import { addStoredAttributes } from '../../properties/addStoredAttributes';
import { addObservers } from '../../properties/addObservers';

export function elementConstructor(self: LSCustomElement, shadow: ShadowOption, otherShadowOptions: OtherShadowOptions) {
  if (shadow) {
    const shadowMode = shadow;
    const attachedShadow = self.attachShadow({ mode: shadowMode, ...otherShadowOptions });
    if (shadowMode === 'closed') {
      self._shadowRoot = attachedShadow;
    }
  }
  self.ls = self.ls || {};
  initStore(self);
  addStores(self);
  addAttributes(self);
  addStoredAttributes(self);
  addObservers(self);
}