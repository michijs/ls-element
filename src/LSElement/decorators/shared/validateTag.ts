export const validateTag = (tag: string) => {
  if (tag.indexOf('-') <= 0) {
    throw new Error('You need at least 1 dash in the custom element name!');
  }
};
// TODO: ver si vale la pena