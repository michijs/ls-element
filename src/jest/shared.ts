import { MainElement } from './MainElement';

export type Result =
    'create1000Rows'
    | 'replaceAllRows'
    | 'partialUpdate'
    | 'selectRow'
    | 'swapRows'
    | 'removeRow'
    | 'createManyRows'
    | 'appendRowsToLargeTable'
    | 'clearRows'

const create1000Rows = () => {
  document.getElementById('run').click();
};
const add1000Rows = () => {
  document.getElementById('add').click();
};
const create10000Rows = () => {
  document.getElementById('runlots').click();
};
const updateEvery10Rows = () => {
  document.getElementById('update').click();
};
const swapRows = () => {
  document.getElementById('swaprows').click();
};
const getTableBody = () => {
  return Array.from(document.getElementById('tbody').children);
};
const clear = () => {
  document.getElementById('clear').click();
};
const getRowId = (element: Element) => {
  return Number(element.childNodes.item(0).childNodes.item(0).textContent);
};

export function makePerformanceTests(dataController: () => MainElement, selectByIndex: boolean) {
  const results = new Map<Result, number>();
  const saveResult = (key: Result, functionToMeasure: () => void) => {
    const t0 = performance.now();
    functionToMeasure();
    const t1 = performance.now();
    results.set(key, Number(Number(t1 - t0).toFixed(2)));
  };
  it('creates 1000 rows when clicking run', () => {
    saveResult('create1000Rows', create1000Rows);
    expect(getTableBody().length).toEqual(1000);
  });
  it('replaces 1000 rows when clicking run', () => {
    create1000Rows();
    saveResult('replaceAllRows', create1000Rows);
    expect(getTableBody().length).toEqual(1000);
    getTableBody().forEach(element => {
      expect(getRowId(element)).toBeGreaterThan(1000);
    });
  });
  it('update every 10th row 1000 rows on a table with 1000 rows when clicking update', () => {
    create1000Rows();
    saveResult('partialUpdate', updateEvery10Rows);
    expect(getTableBody().length).toEqual(1000);
    getTableBody().forEach((element, index) => {
      if (index % 10 === 0) {
        expect(element.innerHTML.includes('!!!')).toBeTruthy();
      } else {
        expect(element.innerHTML.includes('!!!')).toBeFalsy();
      }
    });
  });
  it('select a row (1000 rows)', () => {
    create1000Rows();
    saveResult('selectRow', () => dataController().select(999));
    expect(getTableBody()[selectByIndex ? 999 : 998].className).toEqual('danger');
  });
  it('swap a row (1000 rows)', () => {
    create1000Rows();
    saveResult('swapRows', swapRows);
    expect(getRowId(getTableBody()[1])).toEqual(999);
    expect(getRowId(getTableBody()[998])).toEqual(2);
  });
  it('remove a row (1000 rows)', () => {
    create1000Rows();
    saveResult('removeRow', () => dataController().delete(999));
    expect(getTableBody().length).toEqual(999);
  });
  it('creates 10000 rows when clicking runlots', () => {
    saveResult('createManyRows', create10000Rows);
    expect(getTableBody().length).toEqual(10000);
  });
  it('append 1000 rows on a large table', () => {
    create10000Rows();
    saveResult('appendRowsToLargeTable', add1000Rows);
    expect(getTableBody().length).toEqual(11000);
  });
  it('clear rows', () => {
    create1000Rows();
    saveResult('clearRows', clear);
    expect(getTableBody().length).toEqual(0);
  });
  return results;
}