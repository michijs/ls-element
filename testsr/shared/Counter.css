:host {
  display: flex;
  flex-direction: row;
}

span {
  min-width: 60px;
  text-align: center;
}
