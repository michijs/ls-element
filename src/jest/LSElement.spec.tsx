import 'expect-puppeteer';
import { makePerformanceTests, Result } from './shared';

describe('Performance tests - LS-Element', () => {
  beforeAll(async () => {
    jest.setTimeout(30000);
    await page.goto('http://localhost:3000', {
      waitUntil: 'domcontentloaded'
    });
  });
  afterEach(async () => {
    await page.reload({
      waitUntil: 'domcontentloaded'
    });
  });

  it('matches the snapshot on initial render', async () => {
    const body = await page.$('body');
    const innerHTMLProperty = await body.getProperty('innerHTML');
    const innerHTML = await innerHTMLProperty.jsonValue();
    expect(innerHTML).toMatchSnapshot();
  });

  const results = makePerformanceTests();
  afterAll(async () => {
    expect(
      new Map<Result, number>([
        ['create1000Rows', 628.04],
        ['replaceAllRows', 1338.53],
        ['partialUpdate', 89.95],
        ['selectRow', 61.77],
        ['swapRows', 64.19],
        ['removeRow', 56.07],
        ['createManyRows', 7013.5],
        ['appendRowsToLargeTable', 1579.74],
        ['clearRows', 67.7]
      ])).toMatchSnapshot('LS-Element 1.2.6');
    expect(await results).toMatchSnapshot(new Date().getTime().toString());
  });
});