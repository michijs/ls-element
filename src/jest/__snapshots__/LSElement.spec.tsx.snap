// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Performance tests - LS-Element clear rows: 1625778393550 1`] = `
Map {
  "create1000Rows" => 233.06,
  "replaceAllRows" => 244.17,
  "partialUpdate" => 83.86,
  "selectRow" => 64.31,
  "swapRows" => 78.78,
  "removeRow" => 99.25,
  "createManyRows" => 1825.84,
  "appendRowsToLargeTable" => 502.39,
  "clearRows" => 37.95,
}
`;

exports[`Performance tests - LS-Element clear rows: LS-Element 1.2.6 1`] = `
Map {
  "create1000Rows" => 628.04,
  "replaceAllRows" => 1338.53,
  "partialUpdate" => 89.95,
  "selectRow" => 61.77,
  "swapRows" => 64.19,
  "removeRow" => 56.07,
  "createManyRows" => 7013.5,
  "appendRowsToLargeTable" => 1579.74,
  "clearRows" => 67.7,
}
`;

exports[`Performance tests - LS-Element matches the snapshot on initial render 1`] = `
"
    <div id=\\"main\\">
        <div class=\\"container\\">
            <div class=\\"jumbotron\\">
                <div class=\\"row\\">
                    <div class=\\"col-md-6\\">
                        <h1>LS-Element keyed</h1>
                    </div>
                    <div class=\\"col-md-6\\">
                        <div is=\\"ls-table-manager\\" class=\\"row\\">
                        <div id=\\"6\\" class=\\"col-sm-6 smallpad\\"><button type=\\"button\\" class=\\"btn btn-primary btn-block\\" id=\\"run\\">Create 1,000 rows</button></div><div id=\\"7\\" class=\\"col-sm-6 smallpad\\"><button type=\\"button\\" class=\\"btn btn-primary btn-block\\" id=\\"runlots\\">Create 10,000 rows</button></div><div id=\\"8\\" class=\\"col-sm-6 smallpad\\"><button type=\\"button\\" class=\\"btn btn-primary btn-block\\" id=\\"add\\">Append 1,000 rows</button></div><div id=\\"9\\" class=\\"col-sm-6 smallpad\\"><button type=\\"button\\" class=\\"btn btn-primary btn-block\\" id=\\"update\\">Update every 10th row</button></div><div id=\\"10\\" class=\\"col-sm-6 smallpad\\"><button type=\\"button\\" class=\\"btn btn-primary btn-block\\" id=\\"clear\\">Clear</button></div><div id=\\"11\\" class=\\"col-sm-6 smallpad\\"><button type=\\"button\\" class=\\"btn btn-primary btn-block\\" id=\\"swaprows\\">Swap Rows</button></div></div>
                    </div>
                </div>
            </div>
            <table class=\\"table table-hover table-striped test-data\\">
                <tbody is=\\"ls-table-body\\" id=\\"tbody\\">
                </tbody>
            </table>
            <span class=\\"preloadicon glyphicon glyphicon-remove\\" aria-hidden=\\"true\\"></span>
        </div>
    </div>
    <script src=\\"./index.js\\"></script>


"
`;
